source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -139,137 +139,163 @@ ProcessorState::getCallbackStateAfterHeaderResp(const CommonResponse& common_res
   return ProcessorState::CallbackState::Idle;
 }
 
-// TODO(#37047) Refactoring this function by adding one helper function for each body mode.
 absl::Status ProcessorState::handleHeadersResponse(const HeadersResponse& response) {
-  if (callback_state_ == CallbackState::HeadersCallback) {
-    ENVOY_LOG(debug, "applying headers response. body mode = {}",
-              ProcessingMode::BodySendMode_Name(body_mode_));
-    const auto& common_response = response.response();
-    if (common_response.has_header_mutation()) {
-      const auto mut_status = processHeaderMutation(common_response);
-      if (!mut_status.ok()) {
-        return mut_status;
-      }
+  if (callback_state_ != CallbackState::HeadersCallback) {
+    return absl::FailedPreconditionError("spurious message");
+  }
+
+  ENVOY_LOG(debug, "applying headers response. body mode = {}",
+            ProcessingMode::BodySendMode_Name(body_mode_));
+
+  const auto& common_response = response.response();
+
+  // Process header mutations if present
+  if (common_response.has_header_mutation()) {
+    const auto mut_status = processHeaderMutation(common_response);
+    if (!mut_status.ok()) {
+      return mut_status;
     }
+  }
 
-    clearRouteCache(common_response);
-    onFinishProcessorCall(Grpc::Status::Ok, getCallbackStateAfterHeaderResp(common_response));
+  clearRouteCache(common_response);
+  onFinishProcessorCall(Grpc::Status::Ok, getCallbackStateAfterHeaderResp(common_response));
 
-    if (common_response.status() == CommonResponse::CONTINUE_AND_REPLACE) {
-      ENVOY_LOG(debug, "Replacing complete message");
-      // Completely replace the body that may already exist.
-      if (common_response.has_body_mutation()) {
-        // Remove the content length here because in this case external processor probably won't
-        // properly set the content-length header to match the length of the new body that replaces
-        // the original one.
-        headers_->removeContentLength();
-        body_replaced_ = true;
-        if (bufferedData() == nullptr) {
-          Buffer::OwnedImpl new_body;
-          MutationUtils::applyBodyMutations(common_response.body_mutation(), new_body);
-          addBufferedData(new_body);
-        } else {
-          modifyBufferedData([&common_response](Buffer::Instance& buf) {
-            MutationUtils::applyBodyMutations(common_response.body_mutation(), buf);
-          });
-        }
-      }
+  if (common_response.status() == CommonResponse::CONTINUE_AND_REPLACE) {
+    return handleContinueAndReplace(common_response);
```

Comment:
Thanks for working on this refactoring!

This change seems to be problematic as it skipped the code block:
https://github.com/envoyproxy/envoy/blob/090e73d82462f46d32c112554206a5ed0861eba2/source/extensions/filters/http/ext_proc/processor_state.cc#L266